key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/Verify.java

```diff
@@ -0,0 +1,305 @@
+
+    static {
+        Security.addProvider(new BouncyCastleProvider());
+    }
+
+    /**
+     * Verify given `KeyAttestationResponse` by using `trustRootCa`
+     *
+     * @param keyAttestationResponse Json object in format of
+     *                               `KeyAttestationResponse`, it should contains
+     *                               base64 encoded Authority Certificate chain and
+     *                               Attestation Statement
+     * @param trustRootCa            Trusted root CA, you need to get the
+     *                               certificate
+     *                               from Fortanix PKI website
+     * @param verifyCrl              If check the CRL, need network access
+     * @throws Exception
+     */
+    public static void verify(KeyAttestationResponse keyAttestationResponse,
+            X509Certificate trustRootCa, boolean verifyCrl) throws Exception {
+        LOGGER.info(String.format("Verifying KeyAttestationResponse: %s", keyAttestationResponse.toString()));
+        // skip verifying attestationStatement.format since it's been covered by json
+        // decoding
+        String attestationStatementStr = keyAttestationResponse.getAttestationStatement().getStatement();
+        List<String> authorityChain = keyAttestationResponse.getAuthorityChain();
+        LOGGER.info(
+                String.format("Decoding Attestation Statement Certificate: %s", attestationStatementStr.toString()));
+        X509Certificate attestationStatementCert = readBase64EncodedCertificate(attestationStatementStr);
+        LOGGER.info(String.format("Decoding Authority Chain: %s", authorityChain.toString()));
+        List<X509Certificate> authorityChainCerts = readBase64EncodedCertificates(authorityChain);
+        verify(authorityChainCerts, attestationStatementCert, trustRootCa, verifyCrl);
+    }
+
+    /**
+     * Verify given `attestationStatement`, `authorityChain` by using `trustRootCa`
+     *
+     * @param authorityChain       Certificate chain of all certificates except
+     *                             `Fortanix Key Attestation Statement` certificate
+     * @param attestationStatement `Fortanix Key Attestation Statement` certificate
+     * @param trustRootCa          Trusted root CA, you need to get the certificate
+     *                             from Fortanix PKI website
+     * @param verifyCrl            If check the CRL, need network access
+     * @throws Exception
+     */
+    public static void verify(List<X509Certificate> authorityChain, X509Certificate attestationStatement,
+            X509Certificate trustRootCa, boolean verifyCrl) throws Exception {
+        checkAuthorityChainLength(authorityChain);
+
+        LOGGER.info(String.format("Checking if '%s' certificate is correctly signed by '%s' certificate",
+                Common.DSM_CLUSTER_KEY_ATTESTATION_AUTHORITY_CN, Common.KEY_ATTESTATION_STATEMENT_CN));
```

Comment:
I think the order is still incorrect here.

(This is a nitpick; feel free to ignore)

Reply:
Oh, I see. Eagle eye :+1: